### DISCUSSION TITLE
Using GUI functions?

### DISCUSSION THREAD

**Discussion by afk-Legacy** [vfsfitvnm/frida-il2cpp-bridge#45] [Q&A]
Hey, hope you are doing well. I intend to play around a bit with GUI and wanted to see if I can use the unity functions to create a label of sorts. I am confused as to how to proceed, especially with the first parameter being a struct.

![image](https://user-images.githubusercontent.com/83068093/138282566-9a363801-2e2f-4645-bedd-7b6a5fac7f04.png)

**Comment by vfsfitvnm** [comment]
@afk-Legacy 
Hi, I apologize for the late reply (I don't get notifications from Discussions, uh).
```ts
const RectClass = ...;
const rect = RectClass.alloc();
rect.methods[".ctor"].invoke(1.5, 2.75, 3.0, 4.5);
// convert to struct
const rectAsValueType = rect.unbox();
// ...
Label.invoke(rectAsValueType, ...);
```

**Comment by afk-Legacy** [comment]
@vfsfitvnm 
Hey, I tried above code but it resulted in `expected array with fields` during invocation of Label function.

        const UnityCoreModule = Il2Cpp.Domain.assemblies["UnityEngine.CoreModule"].image;
        const UnityImgui =  Il2Cpp.Domain.assemblies["UnityEngine.IMGUIModule"].image;

        const GUI = UnityImgui.classes["UnityEngine.GUI"];
        const rectClass = UnityCoreModule.classes["UnityEngine.Rect"];

        const rect = rectClass.alloc();
        rect.methods[".ctor"].invoke(0, 30, 100, 20);
        const rectasVal = rect.unbox();
        
        GUI.methods.Label.invoke(rectasVal, Il2Cpp.String.from("Test"));

Resulting in the error:

`Error: expected array with fields
    at NativeFunction.<anonymous> (<anonymous>)
    at ../../node_modules/frida-il2cpp-bridge/dist/il2cpp/structs/method.js:171:1
    at Function.try (../../node_modules/frida-il2cpp-bridge/dist/il2cpp/base.js:116:1)
    at Il2CppMethod.invokeRaw (../../node_modules/frida-il2cpp-bridge/dist/il2cpp/structs/method.js:171:1)
    at Il2CppMethod.invoke (../../node_modules/frida-il2cpp-bridge/dist/il2cpp/structs/method.js:157:1)
    at Il2CppMethod.descriptor.<computed> (../../node_modules/frida-il2cpp-bridge/dist/il2cpp/decorators.js:14:1)
    at index.ts:78:27
    at ../../node_modules/frida-il2cpp-bridge/dist/il2cpp/base.js:95:1`

**Comment by vfsfitvnm** [comment]
@afk-Legacy 
Whops, I forgot `unbox` returns a `NativePointer`, not a `Il2Cpp.ValueType`.
It should be:
```ts
const rectasVal = new Il2Cpp.ValueType(rect.unbox(), rectClass.type);
```

**Reply by afk-Legacy** [reply]
@vfsfitvnm 
Okay so I did what you described but now I am encountering another set of issues, 

1. I seem to have trouble invoking overloaded constructors. (#48)
2. A problem with invocation of another function.

Following your steps, I tried invoking the function but it was failing to invoke the desired constructor and throwing an error, in the meantime, I have tried invoking another function but frida would get stuck at _Attaching...._ when the invoke function was included, without that, it worked fine. I tried invoking the function in the terminal and came across this:

![image](https://user-images.githubusercontent.com/83068093/139593219-4e9ee5b4-8fa3-4eef-b675-bc1ae681d9a4.png)

**Reply by vfsfitvnm** [reply]
I answered at the first one.
Regarding the other one, that is a c# runtime exception, and it's not a frida-il2cpp-bridge issue (I think). Perhaps `GUI.initialize()` may help you

**Reply by afk-Legacy** [reply]
Nope, `GUI.initialize()` also did not work.
